core-v1/src/main/scala/algebra/Expr.scala

```diff
@@ -0,0 +1,355 @@
+
+  implicit def debugAnyExpr[I, O, OP[_]](expr: Expr[I, O, OP]): DebugSyntax[Expr[I, O, OP], Any, Any, OP] =
+    new DebugSyntax(hook => expr.withDebugging(Debugging(hook)))
+
+  implicit def debugExpr[E <: Expr.AnyWith[OP], OP[_]](
+    expr: E,
+  )(implicit
+    debugArgs: DebugArgs[E, OP],
+  ): DebugSyntax[E, debugArgs.In, debugArgs.Out, OP] =
+    new DebugSyntax(debugArgs.attachHook(expr, _))
+
+  /**
+    * Uses the fixed types supplied by the [[DebugArgs]] compiler trick to provide better type inference
+    * when attaching a hook.
+    *
+    * @param attachHook a captured function that will attach a given hook function to the specific [[Expr]] type
+    * @tparam E The specific [[Expr]] subclass used to compute the debug types
+    * @tparam DI The `Debugging Input` type (this doesn't always need to match the actual input to the [[Expr]])
+    *            Typically, this is a tuple of all of the input types of any sub-expressions preceded by the
+    *            input to the given expression.
+    * @tparam DO The `Debugging Output` type (this is always the same as the [[Expr]]'s output type)
+    * @tparam OP The `Output Parameter` (captured at the definition site for every output type in the expression tree)
+    */
+  final class DebugSyntax[E <: Expr.AnyWith[OP], DI, DO, OP[_]](
+    private val attachHook: (ExprState[DI, DO] => Unit) => E,
+  ) extends AnyVal {
+
+    /**
+      * Attach the given function to run when the expression is being interpreted. It allows you to intercept
+      * the computation and inspect the current state, but it does not allow you to alter it.
+      *
+      * TODO: The ability to alter the local state might also be helpful for debugging purposes, but it should
+      *       be something that you can disable for a more "production secure" setup. That said... maybe the
+      *       entire capability of attaching a side-effecting function should be something you can turn off.
+      *       Either way, we should always support this `ExprState => Unit` interface, so it would have to be
+      *       a separate method anyway.
+      */
+    def debug(hook: ExprState[DI, DO] => Unit): E = attachHook(hook)
+  }
+
+  /**
+    * This is the magic that allows you to traverse the entire expression recursively. You need a
+    * visitor type that is contravariant for the final input type and covariant for the final output type.
+    *
+    * The visitor is also parameterized on the "output parameter" `OP[_]`, which is supplied at every node
+    * within the given expression tree. This can be useful information, such as how to serialize the output
+    * of every subexpression, or information about the source code position provided by the compiler, etc.
+    *
+    * Every [[Expr]] subclass will call the associated `visitExpr` method and pass the required constraints.
+    *
+    * @tparam ~> an infix type alias for the higher-kinded result type of this visitor
```

Comment:
Infix type notation is a feature of the Scala language, so I'm not sure how I feel about documenting it here. I am wondering about overloading the `~>` symbol as it is used by cats and shapeless for natural transformations (which is not what this is). I'm thinking maybe `~:>` would be better?

Reply:
_Infix type notation is a feature of the Scala language, so I'm not sure how I feel about documenting it here._ 

I suppose it really depends upon your motivation.  I spoke up about it because I didn't know it.  I usually use one person asking a question as an indication that more people have the same question (or new team members will have) and so clarification is at least necessary.

As to `~>` of `~:>` or just using a letter like we did in the discussion, I am **strongly** in favour of making the code more readable (and less like, say, Perl).